mgmt/mcumgr: Changes in mcumgr source code directory structure

de-nordic opened the pull request:
The PR consists of commits:
 1) Flattening source code tree and changing CMakeLists.txt
 2) ~~moves the smp.h from zephyr/mgmt/mcumgr/smp.h to mgmt/mcumgr/transport/smp.h~~
    All public API interfaces for mcumgr have been moved under include/zephyr/mgmt/mcumgr.
 3) Fixing sample
 4) Fixing tests.
 5) Fixed documentation
 6) Fixes documentation in img_mgmt.h
 7) release notes update

Note that other smp_*.h files have not yet been moved from zephyr/include. This is intentional as currently it is hard to tell whether they define API for SMP transport or provide API to SMP transports from other subsystems.

Build size difference, as tested on nrf52840dk_nrf52840 with command:
```
west build -d smp_svr/ -b nrf52840dk_nrf52840 zephyr/samples/subsys/mgmt/mcumgr/smp_svr/ -DOVERLAY_CONFIG=overlay-bt.conf
```
after changes:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      203012 B       412 KB     48.12%
             RAM:       54796 B       256 KB     20.90%
        IDT_LIST:          0 GB         2 KB      0.00%
```

before changes (01e8f7c0652c9cb9f9ed54a4c9f0d5cd4dd0fd01):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      203336 B       412 KB     48.20%
             RAM:       54796 B       256 KB     20.90%
        IDT_LIST:          0 GB         2 KB      0.00%
```
so we have additional free 324 bytes of flash.

de-nordic committed:
tests/mgmt/mcumgr: Apply rework in MCUMgr subsystem

Code changes required by rework in MCUMgr souce code structure.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
samples/mcumgr/smp_svr: Apply rework in MCUMgr subsystem

Code changes required by rework in MCUMgr souce code structure.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm reviewed (changes requested):
fs_mgmt: hash/checksum registration function, hash/checksum handlers lose visibility
img_mgmt: useful data/structs (to application) lose visibility

de-nordic committed:
doc: Move SMP transport documentation under doc/services

The commit moves MCUmgr SMP transport documentation from
subsys/mgmt/mcumgr/lib to doc/services/device_mgmt.
Documentation have been rewritten for that purpose.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Reworking source code tree and API interface

The commit reworks mgmt/mcumgr subsystem source code to remove
lib subdirectory and make it a little bit more flat.
It also moves all API interface files, which are supposed to be
visible by applications using MCUMgr, to interface sub-directories,
and exposes them with full paths; for example to include general
MCUMgr support, group registration and so on, user would now include:
 <mgmt/mcumgr/mgmt/mgmt.h>
to additionally have control on File System group management
registration user would need:
 <mgmt/mcumgr/mgmt/grp/fs_mgmt.h>

All internal headers have been removed from interface.

CMAkeLists.txt get significant rework and various MCUMgr subsystems
have been divided into separate sub-libraries.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
> fs_mgmt: hash/checksum registration function, hash/checksum handlers lose visibility img_mgmt: useful data/structs (to application) lose visibility

I have changed how include directories are now exposed.
All mcumgr interface headers are directly exposed under paths:
`mgmt/mcumgr/mgmt/` for management API
`mgmt/mcumgr/smp/` for protocol processing
`mgmt/mcumgr/transport/` for transport API
`mgmt/mcumgr/grp/<mgmt_grp>` for groups
where `<mgmt_grp>` are directories like `os_mgmt`, `img_mgmt`, `fs_mgmt` and so on.

All of these include paths also have `priv` directory that should be used for accessing interfaces that should not be exposed out of mcumgr.

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+SMP protocol specification by MCUmgr subsystem of Zephyr uses basic framing
+of data to allow multiplexing of UART channel. Although no such transport
+exists in Zephyr, it is possible to implement MCUmgr client/server over
+uart transport that does not have framing at all, or uses hardware serial
```
*UART, and this is incorrect, SMPc defines that there **must** be frames of no more than 127 bytes each, if this can be allowed to be changed then the SMPc documentation needs to be updated

de-nordic replied:
Quite convinced it is 128 bytes:
https://github.com/zephyrproject-rtos/zephyr/blob/1cb49a389bcfb0c3e5a6ce72cabb61db11fdd6fa/include/zephyr/mgmt/mcumgr/serial.h#L71

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+Frame fragmenting
+=================
+
+SMP protocol over serial is fragmented into MTU size frames; each
```
It's defined as over console: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/transport/smp-console.md
Though I would be fine for updating that to have SMP over serial, either as an addition or replacing the existing over console message

nordicjm replied:
Outstanding comment

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+    +---------------+---------------+---------------------------+
+    | Content       | Size          | Description               |
+    +===============+===============+===========================+
+    | total length  | 2 bytes       | Big endian 16bit value    |
```
```suggestion
    | total length  | 2 bytes       | Big endian 16-bit value   |
```

nordicjm replied:
Unresolved

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+    +===============+===============+===========================+
+    | 0x06 0x09     | 2 bytes       | Frame start marker        |
+    +---------------+---------------+---------------------------+
+    | <base64-i>    | no more than  | Base 64 encoded body      |
```
```suggestion
    | <base64-i>    | no more than  | Base64 encoded body       |
```

nordicjm replied:
Not resolved

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+
+Body of initial frame is preceded by two byte total packet length,
+encoded in big endian, and equals size of raw body plus size of
+two CRC16; this means that actual body size allowed into
```
-two

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/include/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h:
```diff
@@ -8,9 +8,7 @@
 
 #include <inttypes.h>
-#include "img_mgmt_config.h"
-#include "image.h"
```
image.h is needed for `struct image_version` definition

nordicjm replied:
Outstanding comment

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/zephyr_img_mgmt.c:
```diff
@@ -8,20 +8,21 @@
-#include <img_mgmt/img_mgmt.h>
-#include <img_mgmt/image.h>
-#include "img_mgmt_priv.h"
+#include <soc.h>
```
Is this include needed?

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+# OS Management group public API is exposed by MCUMGR API interface,
+# when Statistics Management is enabled.
+add_library(subsys__mgmt__mcumgr__grp__os src/os_mgmt.c)
```
`subsys__mgmt__mcumgr__grp__os` is 29 characters, this is very long, has this been tested in a very long path on windows to ensure it still builds? Can the double underscores be replaced with single underscores?

de-nordic replied:
I was also considering renaming the `subsys__mgmt_mcumgr` as MCUMGR, as it was previously.
I have taken example from another subsystems, but I am not attached to naming, so can rename it. I am open to suggestions here.

nordicjm commented:
Not sure I like this change, it basically breaks every existing code base and requires include updates. zcbor_bulk no longer seems possible to include from an application either.

de-nordic commented:
> Not sure I like this change, it basically breaks every existing code base and requires include updates.

Not a first mcumgr change that does so. Zephyr has already done some header changes (moving stuff under include/zephyr) that also did that.
The change cleans up the tree structure, gives uniform specification for mcumgr header inclusion, and gives better way to avoid aliasing.


> zcbor_bulk no longer seems possible to include from an application either.

This can be fixed. Anyway, I will try to move  `zcbor_bulk` to `zcbor`, as this should be extension over there, rather than in mcumgr.

de-nordic commented:
Also the zcbor_bulk is internal utility for mcumgr and should have not been used outside.

nordicjm commented:
> Also the zcbor_bulk is internal utility for mcumgr and should have not been used outside.

It needs to be usable by applications that add mcumgr functionality. The issue I hit was trying to include it with this PR was in a test.

de-nordic commented:
> > Also the zcbor_bulk is internal utility for mcumgr and should have not been used outside.
> 
> It needs to be usable by applications that add mcumgr functionality. The issue I hit was trying to include it with this PR was in a test.

I will temporarily expose it via MCUMGR interface, although it should not be visible.
Even tests for zcbor_bulk do not pick it via MCUMGR interface:
https://github.com/zephyrproject-rtos/zephyr/blob/caab8cb1b3dc870b564a6ebad567fea4bad1d2ef/tests/subsys/mgmt/mcumgr/zcbor_bulk/CMakeLists.txt#L11-L14

de-nordic commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,197 @@
+for SMP requests: a single-byte characteristic that
+can only accepts write-without-response commands. The contents of
+each write command contains an SMP request.
+
```
Since I am also working on renaming Kconfig, I do not want to that info no, to just rename them in a few weeks.

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+prefixing each frame with two byte marker and terminating it with newline.
+Currently the MCUmgr imposes 128 byte limit on frame size, although there
+are no real protocol contrains that require that limit.
+The limit includes the prefix and the newline character, so a size of
+allowed payload is actually 125 bytes.
```
```suggestion
The limit includes the prefix and the newline character, so the allowed payload
size is actually 125 bytes.
```

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+    +---------------+---------------+---------------------------+
+    | Content       | Size          | Description               |
+    +===============+===============+===========================+
+    | total length  | 2 bytes       | Big endian 16bit value    |
```
```suggestion
    | total length  | 2 bytes       | Big endian 16-bit value   |
```

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+    +===============+===============+===========================+
+    | 0x04 0x14     | 2 bytes       | Frame start marker        |
+    +---------------+---------------+---------------------------+
+    | <base64-i>    | no more than  | Base 64 encoded body      |
```
```suggestion
    | <base64-i>    | no more than  | Base64 encoded body       |
```

nordicjm commented on subsys/mgmt/mcumgr/grp/fs_mgmt/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+
+# File System management group public API is exposed by MCUMGR API
+# interface, when File System management is enabled.
+add_library(MCUMGR__grp__fs src/zephyr_fs_mgmt.c src/fs_mgmt.c)
+
+target_sources_ifdef(CONFIG_FS_MGMT_CHECKSUM_HASH
+	MCUMGR__grp__fs PRIVATE src/hash_checksum_mgmt.c
+)
+target_sources_ifdef(CONFIG_FS_MGMT_CHECKSUM_IEEE_CRC32
+	MCUMGR__grp__fs PRIVATE src/hash_checksum_crc32.c
+)
+target_sources_ifdef(CONFIG_FS_MGMT_HASH_SHA256
+	MCUMGR__grp__fs PRIVATE src/hash_checksum_sha256.c)
```
Follow usage of other in-tree modules with single underscore and lowercase names - for all cmake files.
```
./tests/bluetooth/host/CMakeLists.txt:add_library(host_mocks STATIC
/samples/tfm_integration/tfm_secure_partition/dummy_partition/CMakeLists.txt:add_library(tfm_app_rot_partition_dp STATIC)
./samples/application_development/external_lib/CMakeLists.txt:add_library(mylib_lib STATIC IMPORTED GLOBAL)
./cmake/linker/armlink/target.cmake:    add_library(armlink_steering OBJECT ${STEERING_C})
./cmake/linker/arcmwdt/target.cmake:  add_library(code_relocation_source_lib  STATIC ${MEM_RELOCATION_CODE})
```

de-nordic replied:
Done.

nordicjm commented on subsys/mgmt/mcumgr/grp/fs_mgmt/src/hash_checksum_crc32.c:
```diff
@@ -8,11 +8,10 @@
-#include <fs_mgmt/fs_mgmt_config.h>
-#include <fs_mgmt/fs_mgmt_impl.h>
-#include "fs_mgmt/hash_checksum_mgmt.h"
-#include "fs_mgmt/hash_checksum_crc32.h"
```
add this include back

nordicjm replied:
Not sure why it's showing 4 lines, just the last line for the hash_checksum_crc32.h include

nordicjm replied:
Unresolved

nordicjm commented on subsys/mgmt/mcumgr/grp/fs_mgmt/src/hash_checksum_sha256.c:
```diff
@@ -7,11 +7,9 @@
-#include <fs_mgmt/fs_mgmt_config.h>
-#include <fs_mgmt/fs_mgmt_impl.h>
-#include "fs_mgmt/hash_checksum_mgmt.h"
-#include "fs_mgmt/hash_checksum_sha256.h"
```
add this include back

nordicjm replied:
Unresolved

nordicjm reviewed (changes requested):
zcbor now visible, was able to port test and it worked

de-nordic committed:
mgmt/mcumgr: Move transport headers to transport subdir

The MCUmgr transport headers have been moved to
zephyr/mgmt/mcumgr/transport/

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
mgmt/mcumgr: Correct documentation in img_mgmt.h

Some documentation has not been properly formatted for generating
and some has been inaccurate.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+**************************
+
+SMP communication relays on a single GATT characteristic. An SMP command may
+be sent in the form of either GATT Write command or GATT Write Without Response
+command.
+SMP responses are sent in the form of a GATT Notification, specifying the same
+characteristics as GATT Write command.
```
```suggestion
be sent by using the GATT Write Without Response command.
SMP responses are sent in the form of a GATT Notification, specifying the same
characteristic as the GATT Write Without Response command.
```

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+- **Characteristic UUID**: `DA2E7828-FBCE-4E01-AE9E-261174997C48`
+
+All SMP communication utilizes a single GATT characteristic.  An SMP request is
+sent in the form of either 1) a GATT Write Command, or 2) a GATT Write Without
```
```suggestion
sent via a GATT Write Without Response command. An SMP response is sent in the form
of a GATT Notification
```

nordicjm replied:
And this information is duplicated above already

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+Response command. An SMP response is sent in the form of a GATT Notification
+specifying the same characteristic that was written.
+
+The Write Without Response form is generally preferred, as an application-layer
+response is always sent in the form of a Notification. The regular Write form is
+accepted in case the client requires a GATT response to initiate pairing.
```
It is not

de-nordic replied:
Remove the last line.

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+accepted in case the client requires a GATT response to initiate pairing.
+
+If an SMP request or response is too large to fit in a single GATT command, the
+sender fragments it across several commands.  No additional framing is
```
```suggestion
sender fragments it across several packets.  No additional framing is
```
All of them would be a single command, thus multiple packets for a single command/response

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+If an SMP request or response is too large to fit in a single GATT command, the
+sender fragments it across several commands.  No additional framing is
+introduced when a request or response is fragmented; the payload is simply
+split among several packets.  Since Bluetooth guarantees ordered delivery of
```
Note that GATT guarantees delivery order, EATT does not, I'm assuming that EATT does not just replace GATT though and would need to change the whole of how an application works so can be ignored

de-nordic replied:
Changed sentence to `Since GATT`

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+packets, the SMP header in the first fragment contains sufficient information
+for reassembly.
+
+Note: Zephyr currently supports only GATT Write Without Response, and sending
+responses in form of GATT Notifications.
```
This part can go

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+SMP protocol specification by MCUmgr subsystem of Zephyr uses basic framing
+of data to allow multiplexing of UART channel. Multiplexing requires
+prefixing each frame with two byte marker and terminating it with newline.
+Currently MCUmgr imposes a 128 byte limit on frame size, although there
```
```suggestion
Currently MCUmgr imposes a 127 byte limit on frame size, although there
```

nordicjm replied:
This is wrong then, header is 3 bytes, assuming 128 byte packet, that leaves 125 bytes for data, the data is in base64, 125 / 4 = 31.25, 31.25 * 3 = 93.75 - this is not valid, therefore it allows 93 base64 encoded bytes, 93 base64 bytes is occupies 124 bytes of data, 124 + 3 = 127, so the maximum is 127, not 128.

nordicjm replied:
Unresolved

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+of data to allow multiplexing of UART channel. Multiplexing requires
+prefixing each frame with two byte marker and terminating it with newline.
+Currently MCUmgr imposes a 128 byte limit on frame size, although there
+are no real protocol contrains that require that limit.
```
```suggestion
are no real protocol constraints that require that limit.
```

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+Currently MCUmgr imposes a 128 byte limit on frame size, although there
+are no real protocol contrains that require that limit.
+The limit includes the prefix and the newline character, so the allowed payload
+size is actually 125 bytes.
```
```suggestion
size is actually 124 bytes.
```

de-nordic replied:
128 - 3 = 125:
https://github.com/zephyrproject-rtos/zephyr/blob/1cb49a389bcfb0c3e5a6ce72cabb61db11fdd6fa/include/zephyr/mgmt/mcumgr/serial.h#L71

nordicjm replied:
Unresolved, with base64, the amount of data is 124, not 125

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,214 @@
+frame consists of two byte start marker, body and terminating newline
+character.
+Note: currently the MTU for UART with this fragmenting scheme is
+limited to 128 bytes, and includes start marker and terminating newline.
```
This information is duplicated from above, can be removed

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# OS Management group public API is exposed by MCUmgr API interface,
+# when Statistics Management is enabled.
```
??

de-nordic replied:
Fixed.

nordicjm commented on subsys/mgmt/mcumgr/grp/zephyr_basic/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Statistics management group public API is exposed by MCUmgr API
+# interface, when Statistics management is enabled.
```
Don't think this is relevant to this file

de-nordic replied:
Yes, fixed.

nordicjm commented on subsys/mgmt/mcumgr/mgmt/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# Main MCUmgr interface source and API.
+# Provides MCUmgr services, group reistration
```
```suggestion
# Provides MCUmgr services, group registration, event handling etc.
```

nordicjm replied:
yet another github bug, line 7 should not be included in this suggestion

nordicjm commented on subsys/mgmt/mcumgr/smp/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Protocol API is only exposed to MCUMGR internals.
```
```suggestion
# Protocol API is only exposed to MCUmgr internals.
```

nordicjm replied:
yet another github bug, line 7 should not be included in this suggestion

nordicjm commented on subsys/mgmt/mcumgr/util/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# MCUmgr utilities, for use within the library.
+# API interface for utilities is exposed by mgmt_mcumgr_util,
+# and should not be exposed outside of mgmt_mcumgr.
+add_library(mgmt_mcumgr_util src/zcbor_bulk.c)
```
Need a task to move zcbor_bulk over to zcbor itself and remove the implementation from mcumgr

de-nordic replied:
Working on it.

nordicjm commented:
@de-nordic needs rebase

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,208 @@
+BLE (Bluetooth Low Energy)
+**************************
+
+SMP communication relays on a single GATT characteristic. An SMP command may
+be sent by using the GATT Write Without Response command.
+SMP responses are sent in the form of a GATT Notification, specifying the same
+characteristic as the GATT Write Without Response command.
+
+BLE (Bluetooth Low Energy) Characteristics
+==========================================
+
+MCUmgr Clients need to use following BLE Characteristics, when implementing
+SMP client:
+
+- **Service UUID**: `8D53DC1D-1DB7-4CD3-868B-8A527460AA84`
+- **Characteristic UUID**: `DA2E7828-FBCE-4E01-AE9E-261174997C48`
+
+All SMP communication utilizes a single GATT characteristic.  An SMP request is
+sent via a GATT Write Without Response command. An SMP response is sent in the form
+of a GATT Notification
+Response command. An SMP response is sent in the form of a GATT Notification
+specifying the same characteristic that was written.
```
One of these needs to go, they're saying the exact same thing

nordicjm commented on doc/services/device_mgmt/smp_transport.rst:
```diff
@@ -0,0 +1,208 @@
+Response command. An SMP response is sent in the form of a GATT Notification
+specifying the same characteristic that was written.
+
+The Write Without Response form is generally preferred, as an application-layer
+response is always sent in the form of a Notification.
```
Not valid, it is the only method, so this can go

de-nordic replied:
Gone.

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h:
```diff
@@ -105,89 +102,107 @@ struct img_mgmt_upload_action {
- * @brief Read info of an image give the slot number
+ * @brief Read info of an image at the specified slot number
+ *
+ * @param image_slot	 image slot number;
+ * @param ver		 output buffer for image version;
+ * @param hash		 output buffer for image hash;
+ * @param flags		 output buffer for image flags,
```
remove ; and , from end of lines, this is not in line with zephyr documentation

de-nordic replied:
Done.

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h:
```diff
@@ -105,89 +102,107 @@ struct img_mgmt_upload_action {
- *					  information
- * @param dst		   Destination string created from the given
- *					  in image version
+ * @param ver		pointer to image_version object;
```
As above, remove ;

de-nordic replied:
Done.

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h:
```diff
@@ -105,89 +102,107 @@ struct img_mgmt_upload_action {
  *
- * @param slot	   Image slot to set pending
- * @param permanent  If set no confirm is required after image swap
+ * @param slot		slot number;
```
Remove ;

de-nordic replied:
Done.

de-nordic committed:
doc/release/3.3: Notes on changes in MCUmgr subsystem tree

Added notes on changes in MCUmgr subsystem tree.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
Does anyone know better tool for github reviews? I am getting lost in comments here...

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -287,6 +287,37 @@ Libraries / Subsystems
+       +----------------+-------------------------------------------+
+
+    Public API interfaces for above areas are now exported through zephyr_interface.
+    Public API interfaces for above areas are now exported through zephyr_interfae,
```
Duplicare

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -287,6 +287,37 @@ Libraries / Subsystems
+    Public API interfaces for above areas are now exported through zephyr_interface.
+    Public API interfaces for above areas are now exported through zephyr_interfae,
+    and headers for them reside in ``zephyr/mgmt/mcumgr/<mcumgr_subarea>/``.
+    For example to acces mgmt API include ``<zephyr/mgmt/mcumgr/mgmt/mgmt.h>``.
```
```suggestion
    For example to access mgmt API include ``<zephyr/mgmt/mcumgr/mgmt/mgmt.h>``.
```

de-nordic committed:
samples/nrf/onoff_level_light: Convert to new MCUmgr header paths

Header paths changed.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on doc/releases/release-notes-3.3.rst:
```diff
@@ -327,6 +327,36 @@ Libraries / Subsystems
+       |                | MCUmgr.                                   |
+       +----------------+-------------------------------------------+
+
+    Public API interfaces for above areas are now exported through zephyr_interfae,
```
```suggestion
    Public API interfaces for above areas are now exported through zephyr_interface,
```

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h:
```diff
@@ -9,14 +9,10 @@
-#include "img_mgmt_config.h"
-#include "image.h"
-#include "mgmt/mgmt.h"
-#include "smp/smp.h"
```
smp.h needed for `struct smp_streamer` definition

de-nordic replied:
Base64 or not I am quite sure this is 128 characters.

nordicjm replied:
Would fail the checks, it would treat it as part of the base64 decode which would then error out. This is a binary protocol, not a text file, so should never be sent with windows-style newlines.

de-nordic replied:
Fixed with revase over #52221 merge.

nordicjm commented:
@de-nordic needs rebase

nordicjm commented:
Test failures need fixing but changes look good

nordicjm commented:
@de-nordic needs rebase

de-nordic commented:
> @de-nordic needs rebase

Done
